File: Liftr/Pet/PetCombatPlaybackEngine.swift
import Foundation
import Observation

@MainActor
@Observable
final class PetCombatPlaybackEngine {
    private(set) var currentTurnIndex: Int = -1
    private(set) var currentMessage: String = ""
    private(set) var isPlaying = false
    private(set) var isFinished = false
    private(set) var hitSide: PetCombatHitSide?
    private(set) var lastStrike: PetCombatStrikeEvent?
    private(set) var attackerCurrentHp: Int
    private(set) var defenderCurrentHp: Int
    let attackerMaxHp: Int
    let defenderMaxHp: Int

    enum PetCombatHitSide {
        case attacker
        case defender
    }

    struct PetCombatStrikeEvent: Equatable, Identifiable {
        let id: Int
        let target: PetCombatHitSide
        let damage: Int
        let isCritical: Bool
        let isDodged: Bool
    }

    private let turns: [PetBattleTurn]
    private let turnDelayNanoseconds: UInt64

    init(
        turns: [PetBattleTurn],
        attackerMaxHp: Int,
        defenderMaxHp: Int,
        turnDelayMilliseconds: Int = 1000
    ) {
        self.turns = turns
        self.attackerMaxHp = max(attackerMaxHp, 1)
        self.defenderMaxHp = max(defenderMaxHp, 1)
        self.attackerCurrentHp = self.attackerMaxHp
        self.defenderCurrentHp = self.defenderMaxHp
        self.turnDelayNanoseconds = UInt64(turnDelayMilliseconds) * 1_000_000
    }

    func start() async {
        guard !turns.isEmpty else {
            isFinished = true
            return
        }

        isPlaying = true
        isFinished = false
        currentTurnIndex = -1
        currentMessage = ""
        attackerCurrentHp = attackerMaxHp
        defenderCurrentHp = defenderMaxHp

        for (index, turn) in turns.enumerated() {
            try? await Task.sleep(nanoseconds: turnDelayNanoseconds)
            currentTurnIndex = index
            currentMessage = turn.message
            attackerCurrentHp = turn.attackerHpAfter
            defenderCurrentHp = turn.defenderHpAfter
            let isDodged = turn.action == "dodge"
            let target: PetCombatHitSide = turn.actor == "attacker" ? .defender : .attacker
            lastStrike = PetCombatStrikeEvent(
                id: index,
                target: target,
                damage: turn.damage,
                isCritical: turn.isCritical,
                isDodged: isDodged
            )
            if !isDodged {
                hitSide = target
            }
            try? await Task.sleep(nanoseconds: 250_000_000)
            hitSide = nil
        }

        isPlaying = false
        isFinished = true
    }
}

File: android/app/src/main/java/com/lilru/liftr/ui/pets/PetCombatPlaybackEngine.kt
package com.lilru.liftr.ui.pets

import com.lilru.liftr.data.PetBattleTurnWire
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

enum class PetCombatHitSide {
    Attacker,
    Defender
}

data class PetCombatStrikeEvent(
    val id: Int,
    val target: PetCombatHitSide,
    val damage: Int,
    val isCritical: Boolean,
    val isDodged: Boolean
)

class PetCombatPlaybackEngine(
    private val turns: List<PetBattleTurnWire>,
    attackerMaxHp: Int,
    defenderMaxHp: Int,
    private val turnDelayMs: Long = 1000L
) {
    val attackerMaxHp: Int = maxOf(attackerMaxHp, 1)
    val defenderMaxHp: Int = maxOf(defenderMaxHp, 1)

    private val _currentTurnIndex = MutableStateFlow(-1)
    val currentTurnIndex: StateFlow<Int> = _currentTurnIndex.asStateFlow()

    private val _currentMessage = MutableStateFlow("")
    val currentMessage: StateFlow<String> = _currentMessage.asStateFlow()

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _isFinished = MutableStateFlow(false)
    val isFinished: StateFlow<Boolean> = _isFinished.asStateFlow()

    private val _hitSide = MutableStateFlow<PetCombatHitSide?>(null)
    val hitSide: StateFlow<PetCombatHitSide?> = _hitSide.asStateFlow()

    private val _lastStrike = MutableStateFlow<PetCombatStrikeEvent?>(null)
    val lastStrike: StateFlow<PetCombatStrikeEvent?> = _lastStrike.asStateFlow()

    private val _attackerCurrentHp = MutableStateFlow(this.attackerMaxHp)
    val attackerCurrentHp: StateFlow<Int> = _attackerCurrentHp.asStateFlow()

    private val _defenderCurrentHp = MutableStateFlow(this.defenderMaxHp)
    val defenderCurrentHp: StateFlow<Int> = _defenderCurrentHp.asStateFlow()

    suspend fun start() {
        if (turns.isEmpty()) {
            _isFinished.value = true
            return
        }
        _isPlaying.value = true
        _isFinished.value = false
        _currentTurnIndex.value = -1
        _currentMessage.value = ""
        _attackerCurrentHp.value = attackerMaxHp
        _defenderCurrentHp.value = defenderMaxHp

        turns.forEachIndexed { index, turn ->
            delay(turnDelayMs)
            _currentTurnIndex.value = index
            _currentMessage.value = turn.message
            _attackerCurrentHp.value = turn.attackerHpAfter
            _defenderCurrentHp.value = turn.defenderHpAfter
            val isDodged = turn.action == "dodge"
            val target = if (turn.actor == "attacker") PetCombatHitSide.Defender else PetCombatHitSide.Attacker
            _lastStrike.value = PetCombatStrikeEvent(
                id = index,
                target = target,
                damage = turn.damage,
                isCritical = turn.isCritical,
                isDodged = isDodged
            )
            if (!isDodged) {
                _hitSide.value = target
            }
            delay(250)
            _hitSide.value = null
        }

        _isPlaying.value = false
        _isFinished.value = true
    }
}
